### PR TITLE
CLOUDNS: add support for LOC records

### PR DESCRIPTION
**Pull request by hmoffatt**
This PR adds support for NAPTR and LOC records on CLOUDNS.

### CODE REVIEW

**Review by tlimoncelli — APPROVED** [review]

**Comment by tlimoncelli** [comment]
Thanks!

**Review by tlimoncelli — APPROVED** [review]

**Comment by tlimoncelli** [comment]
Please add a comment with the output of the integration tests (redact as needed)

File: providers/cloudns/cloudnsProvider.go
Comment (systemcrash) [inline providers/cloudns/cloudnsProvider.go:347]: This part should be redundant now.

File: providers/cloudns/cloudnsProvider.go
Comment (systemcrash) [inline providers/cloudns/cloudnsProvider.go:429]: And this.

**Comment by hmoffatt** [comment]
Oops, I forgot the unit tests. This showed up an interesting quirk with CLOUDNS's handling of the size/precision fields - it quietly substitutes its own default values if you pass "0.00", even though those are documented as valid! I've put in a workaround.

File: providers/cloudns/cloudnsProvider.go
Comment (hmoffatt) [reply providers/cloudns/cloudnsProvider.go:347]: I think specifically the part that strips the decimal part of altitude is redundant? The rest has to remain as I need to assemble the LOC record from the individual API parts.

Better to call `calculateLOCFields` directly, but then I would have to duplicate the string to int/float conversions as the API returns strings for everything, despite being JSON.

File: providers/cloudns/cloudnsProvider.go
Comment (hmoffatt) [reply providers/cloudns/cloudnsProvider.go:429]: I'll wait for your change to be merged then rebase this. Thanks.

**Comment by hmoffatt** [comment]
All tests are passing now except the NAPTR test. The CLOUDNS API is rejecting records with both `regexp` and `replacement` set, which the test cases use. I couldn't find anywhere in the RFC that explicitly said this wasn't allowed, but none of the examples use both fields.

```
=== RUN   TestDNSProviders/dnscontrol-test.dev/35:NAPTR:NAPTR_record
    integration_test.go:248: 
        + CREATE NAPTR test.dnscontrol-test.dev 100 10 "U" "E2U+sip" "!^.*$!sip:customer-service@example.com!" example.foo.com. ttl=300
    integration_test.go:253: failed create record (ClouDNS): ClouDNS API error: You need to enter either Regular Expression or Replacement, not both
```

File: providers/cloudns/cloudnsProvider.go
Comment (systemcrash) [reply providers/cloudns/cloudnsProvider.go:347]: > Better to call `calculateLOCFields` directly, but then I would have to duplicate the string to int/float conversions as the API returns strings for everything, despite being JSON.

🤦 

I don't know what to say. But coercing as early as possible is safest, especially when working with numbers and maths.

**Comment by systemcrash** [comment]
> Oops, I forgot the unit tests. This showed up an interesting quirk with CLOUDNS's handling of the size/precision fields - it quietly substitutes its own default values if you pass "0.00", even though those are documented as valid! I've put in a workaround.

Can you supply nothing?

`helpers.js` assumes 0 if the data source is DMS or DMM using a domain.js. But no assumptions or changes are made for records already deployed in a domain.

**Comment by systemcrash** [comment]
> All tests are passing now except the NAPTR test. The CLOUDNS API is rejecting records with both `regexp` and `replacement` set, which the test cases use. I couldn't find anywhere in the RFC that explicitly said this wasn't allowed, but none of the examples use both fields.
> 
> ```
> === RUN   TestDNSProviders/dnscontrol-test.dev/35:NAPTR:NAPTR_record
>     integration_test.go:248: 
>         + CREATE NAPTR test.dnscontrol-test.dev 100 10 "U" "E2U+sip" "!^.*$!sip:customer-service@example.com!" example.foo.com. ttl=300
>     integration_test.go:253: failed create record (ClouDNS): ClouDNS API error: You need to enter either Regular Expression or Replacement, not both
> ```

Seems pretty wild. Both are necessary for a functioning NAPTR pointer. Flag an error with their support. 

Recommend that you separate out NAPTR changes to a separate PR if you think this could take a while to resolve.

**Comment by hmoffatt** [comment]
> Can you supply nothing?

It turns out that the weird behaviour with "0.00" also applies to the altitude field. If I supply nothing, I get the default of 10,000m. :confused:

**Comment by hmoffatt** [comment]
> Recommend that you separate out NAPTR changes to a separate PR if you think this could take a while to resolve.

Done. Once your #3130 is merged I will submit this PR.

**Comment by hmoffatt** [comment]
I've stripped out the NAPTR code for now, and updated this for @systemcrash 's decimal LOC changes. 
[cloudns-test.log](https://github.com/user-attachments/files/17171453/cloudns-test.log)

All the unit tests are passing, so I think this is ready to go! Thanks all for review and assistance.

**Comment by systemcrash** [comment]
Sorry lads, I've no experience with cloud but your approach looks sound. 🤓

**Comment by hmoffatt** [comment]
Hi @tlimoncelli I think this is ready to go now. Thanks.

**Comment by tlimoncelli** [comment]
Thanks!

**Comment by systemcrash** [comment]
@hmoffatt have your changes proved beneficial?

**Comment by hmoffatt** [comment]
My interest in this was academic tbh, just learning my way around Go and the dnscontrol codebase.

**Pull request merged**
